lib/std/sums.nim

```diff
@@ -53,13 +53,111 @@ func sumPairs*[T](x: openArray[T]): T =
   let n = len(x)
   if n == 0: T(0) else: sumPairwise(x, 0, n)
 
+func fastTwoSum*[T: SomeFloat](a, b: T): (T, T) =
+  ## Deker's algorithm
+  ## pre-condition: |a| >= |b|
+  ## you must swap a and b if pre-condition is not satisfied
+  ## s + r = a + b exactly (s is result[0] and r is result[1])
+  ## s is the nearest FP number of a + B
+  assert(abs(a) >= abs(b))
+  result[0] = a + b
+  result[1] = b - (result[0] - a)
+
+func twoSum*[T](a, b: T): (T, T) =
+  ## Møller-Knuth's algorithm.
+  ## Improve Deker's algorithm, no branch needed
+  ## More operations, but still cheap.
+  result[0] = a + b
+  let z = result[0] - a
+  result[1] = (a - (result[0] - z)) + (b - z)
+
+func sum2*[T: SomeFloat](v: openArray[T]): T =
+  ## sum an array v using twoSum function
+  if len(v) == 0:
+    return 0.0
+
+  var s = v[0]
+  var e: float
+  for i in 1..<v.len-1:
+    let sum = twoSum(s, v[i])
+    s = sum[0]
+    e += sum[1]
+  return s + e
+
+func sumShewchuck_add[T: SomeFloat](v: openArray[T]): seq[T] =
+  for x in v:
```

Comment:
i was concerned about the fact this allocates and could be inefficient but in my limited tests, 2 was the max size for result.len.
maybe add a comment to explain this